File: src/main/java/at/searles/parsing/combinators/ParserThenReducer.java
package at.searles.parsing.combinators;

import at.searles.parsing.*;
import at.searles.parsing.printing.ConcreteSyntaxTree;
import at.searles.parsing.printing.PartialConcreteSyntaxTree;
import at.searles.parsing.printing.PrinterBacktrackException;

/**
 * Parser for chaining parsers. Eg for 5 + 6 where + 6 is the reducer.
 */
public class ParserThenReducer<T, U> implements Parser<U>, Recognizable.Then {

    private final Parser<T> parent;
    private final Reducer<T, U> reducer;

    private final boolean allowParserBacktrack;
    private final boolean allowPrinterBacktrack;

    public ParserThenReducer(Parser<T> parent, Reducer<T, U> reducer, boolean allowParserBacktrack, boolean allowPrinterBacktrack) {
        this.parent = parent;
        this.reducer = reducer;
        this.allowParserBacktrack = allowParserBacktrack;
        this.allowPrinterBacktrack = allowPrinterBacktrack;
    }

    @Override
    public U parse(ParserStream stream) {
        long offset = stream.getOffset();

        // to restore if backtracking
        long preStart = stream.getStart();
        long preEnd = stream.getEnd();

        T t = parent.parse(stream);

        if (t == null) {
            return null;
        }

        // reducer preserves start() in stream and only sets end().
        U u = reducer.parse(stream, t);

        if (u == null) {
            if(offset != stream.getOffset()) {
                throwIfNoBacktrack(stream);
                stream.backtrackToOffset(offset);
            }

            stream.setStart(preStart);
            stream.setEnd(preEnd);

            return null;
        }

        return u;
    }

    @Override
    public Recognizable left() {
        return parent;
    }

    @Override
    public Recognizable right() {
        return reducer;
    }

    @Override
    public boolean allowParserBacktrack() {
        return allowParserBacktrack;
    }

    @Override
    public ConcreteSyntaxTree print(U u) {
        PartialConcreteSyntaxTree<T> reducerOutput = reducer.print(u);

        if (reducerOutput == null) {
            return null;
        }

        ConcreteSyntaxTree parserOutput = parent.print(reducerOutput.left);

        if (parserOutput == null) {
            if(!allowPrinterBacktrack) {
                throw new PrinterBacktrackException(this, reducerOutput.right);
            }

            return null;
        }

        return parserOutput.consRight(reducerOutput.right);
    }

    @Override
    public String toString() {
        return createString();
    }
}

File: src/main/java/at/searles/parsing/combinators/ParserThenReducer.kt
package at.searles.parsing.combinators

import at.searles.parsing.*
import at.searles.parsing.printing.ConcreteSyntaxTree

/**
 * Parser for chaining parsers. Eg for 5 + 6 where + 6 is the reducer.
 */
class ParserThenReducer<T, U>(private val left: Parser<T>, private val right: Reducer<T, U>) : Parser<U> {
    override fun parse(stream: ParserStream): U? {
        return stream.parse(left, true)?.let {
            stream.reduce(it, right)
        }
    }

    override fun recognize(stream: ParserStream): Boolean {
        return stream.recognize(left, true) && stream.recognize(right, false)
    }

    override fun print(item: U): ConcreteSyntaxTree? {
        val reducerOutput = right.print(item) ?: return null
        val parserOutput = left.print(reducerOutput.left) ?: return null
        return parserOutput.consRight(reducerOutput.right)
    }

    override fun <V> plus(right: Reducer<U, V>): Parser<V> {
        return ParserThenReducer(left, this.right + right)
    }

    override fun toString(): String {
        return "$left.then($right)"
    }
}